core-bundle/src/EventListener/DataContainer/UndoListener.php

```diff
@@ -0,0 +1,347 @@
+    private ContaoFramework $framework;
+
+    /**
+     * @var Controller
+     */
+    private $controller;
+
+    /**
+     * @var Backend
+     */
+    private $backend;
+
+    /**
+     * @var Image
+     */
+    private $image;
+
+    public function __construct(ContaoFramework $framework, Connection $connection)
+    {
+        $this->framework = $framework;
+        $this->connection = $connection;
+
+        /** @var Controller $controller */
+        $controller = $framework->getAdapter(Controller::class);
+        $this->controller = $controller;
+
+        /** @var Backend $backend */
+        $backend = $framework->getAdapter(Backend::class);
+        $this->backend = $backend;
+
+        /** @var Image $image */
+        $image = $this->framework->getAdapter(Image::class);
+        $this->image = $image;
+    }
+
+    /**
+     * @Callback(target="list.label.label", table="tl_undo")
+     */
+    public function renderUndoLabel(array $row, string $label, DataContainer $dc): string
+    {
+        $table = $row['fromTable'];
+        $data = StringUtil::deserialize($row['data']);
+        $originalRow = $data[$table][0];
+
+        $this->controller->loadDataContainer($table);
+        $this->controller->loadLanguageFile($table);
+
+        $dataContainer = $this->framework->getAdapter(DataContainer::class)->getDriverForTable($table);
+        $dc = new $dataContainer($table);
+
+        $header = $this->renderHeader($table, $row, $originalRow);
+        $label = $this->renderLabel($originalRow, $dc);
+
+        if ($GLOBALS['TL_DCA'][$table]['list']['label']['showColumns'] ?? false) {
+            $label = $this->renderColumns($label);
+        }
+
+        return $header.'<div class="tl_undo_preview">'.$label.'</div>';
+    }
+
+    /**
+     * @Callback(target="list.operations.jumpToParent.button", table="tl_undo")
+     */
+    public function renderJumpToParentButton(array $row, ?string $href = '', string $label = '', string $title = '', string $icon = '', string $attributes = ''): string
+    {
+        $data = StringUtil::deserialize($row['data']);
+        $table = $row['fromTable'];
+        $originalRow = $data[$table][0];
+        $type = $this->getTypeFromTable($table);
+        $title = sprintf($GLOBALS['TL_LANG']['tl_undo']['parent_modal'], $type, $originalRow['id']);
+        $parent = $this->getParentTableForRow($table, $originalRow);
+
+        if (!$parent || !$this->checkIfParentExists($parent)) {
+            return $this->image->getHtml('parent_.svg', $label).' ';
+        }
+
+        $params = $this->getParentLinkParameters($parent, $table);
+
+        return sprintf(
+            '<a href="%s" title="%s" onclick="Backend.openModalIframe({\'title\':\'%s\',\'url\': this.href });return false">%s</a> ',
+            $this->backend->addToUrl($params),
+            StringUtil::specialchars($title),
+            $title,
+            $this->image->getHtml($icon, $label)
+        );
```

Comment:
Two remarks that do not specifically apply to this method:

1.  If you move variable declarations down to where they are needed (basically as far down as possible), they often do not need to be computed in the first place if an early return or guard is hit. And if they are used only once, you can easily inline them.

2. Please make sure not to overwrite arguments with same name (`$title` in this case). If you're using PHPStorm the "PHP Inspections (EA Extended)" plugin will easily help you find those things. The inspection is called "Suspicious assignments" and is enabled by default.

With this in mind, here is how this could get refactored:

```suggestion
        $table = $row['fromTable'];
        $originalRow = StringUtil::deserialize($row['data'])[$table][0];
        $parent = $this->getParentTableForRow($table, $originalRow);

        if (!$parent || !$this->checkIfParentExists($parent)) {
            return $this->image->getHtml('parent_.svg', $label).' ';
        }

        $newTitle = sprintf(
            $GLOBALS['TL_LANG']['tl_undo']['parent_modal'], 
            $this->getTypeFromTable($table), 
            $originalRow['id']
        );

        return sprintf(
            '<a href="%s" title="%s" onclick="Backend.openModalIframe({\'title\':\'%s\',\'url\': this.href });return false">%s</a> ',
            $this->backend->addToUrl($this->getParentLinkParameters($parent, $table)),
            StringUtil::specialchars($newTitle),
            $newTitle,
            $this->image->getHtml($icon, $label)
        );
```